### PR TITLE
feat: add Pipeline Parallelism (PP) and PD  support for DeepSeek-V4

### PR DESCRIPTION
**Pull request by whybeyoung**
Key changes for PP support with mHC (Manifold-Constrained Hyper-Connections):

1. deepseek_v4.py:
   - PP rank guards: skip embed_tokens/hc_head/norm/logits on non-applicable ranks
   - Flatten 3D mHC hidden_states (seq, hc_mult, hidden) to 2D for PP send, unflatten back to 3D on receive
   - Skip PPMissingLayer in post_load_weights and _setup_fp8_wo_a_scales
   - Expose start_layer/end_layer on ForCausalLM for model_runner PP support

2. deepseekv4_memory_pool.py:
   - PP stage-aware layer mapping (_stage_start/_stage_end)
   - Add _swa_layer_id() to convert absolute layer_id to pool-local index
   - Fix cached_loc invalidation for PP (start_layer check)

3. scheduler_pp_mixin.py:
   - Use spec_hidden_size for PP proxy tensor allocation (mHC needs hc_mult * hidden_size instead of hidden_size)

4. cuda_graph_runner.py / input_buffers.py:
   - Pass pp_hidden_size (spec_hidden_size) for CUDA graph PP proxy tensors
   - hidden_states buffer uses spec_hidden_size, residual stays at hidden_size

## DeepSeek v4 FLash
with disable cudagraph:
```
 SGLANG_DSV4_FP4_EXPERTS=0 \
 SGLANG_JIT_DEEPGEMM_PRECOMPILE=0 \
nohup  python -m sglang.launch_server \
   --trust-remote-code \
   --mem-fraction-static 0.8 \
   --model-path /work/models \
   --tp-size 8 \
   --pp-size 2\
   --dp-size 1 \
   --enable-dp-attention \
   --host 0.0.0.0 \
   --port 30000 \
   --nnodes 2 \
   --node-rank 0 \
   --disable-cuda-graph \
   --dist-init-addr 26.5.27.239:20000 \
   --max-running-requests 256  &
   
   
 SGLANG_DSV4_FP4_EXPERTS=0 \
 SGLANG_JIT_DEEPGEMM_PRECOMPILE=0 \
 nohup python -m sglang.launch_server \
   --trust-remote-code \
   --mem-fraction-static 0.8 \
   --model-path /work/models \
   --tp-size 8 \
   --pp-size 2\
   --dp-size 1 \
   --enable-dp-attention \
   --nnodes 2 \
   --node-rank 1 \
     --disable-cuda-graph \
   --dist-init-addr 26.5.27.239:20000 \
   --max-running-requests 256  &
   
```
gsm 8k:
image

with cuda graph:
```
SGLANG_DSV4_FP4_EXPERTS=0 \
 SGLANG_JIT_DEEPGEMM_PRECOMPILE=0  \
nohup  python -m sglang.launch_server \
   --trust-remote-code \
   --mem-fraction-static 0.8 \
   --model-path /work/models \
   --tp-size 8 \
   --pp-size 2\
   --dp-size 1 \
   --enable-dp-attention \
   --host 0.0.0.0 \
   --port 30000 \
   --nnodes 2 \
   --node-rank 0 \
   --dist-init-addr 26.5.27.239:20000 \
   --max-running-requests 256  &
   
   
 SGLANG_DSV4_FP4_EXPERTS=0 \
 SGLANG_JIT_DEEPGEMM_PRECOMPILE=0 \
 nohup python -m sglang.launch_server \
   --trust-remote-code \
   --mem-fraction-static 0.8 \
   --model-path /work/models \
   --tp-size 8 \
   --pp-size 2\
   --dp-size 1 \
   --enable-dp-attention \
   --nnodes 2 \
   --node-rank 1 \
   --dist-init-addr 26.5.27.239:20000 \
   --max-running-requests 256  &
```
gsm8k:
image

## For PD

prefill:
```
 SGLANG_DSV4_FP4_EXPERTS=0 
 SGLANG_JIT_DEEPGEMM_PRECOMPILE=0 
 nohup python -m sglang.launch_server \
   --trust-remote-code \
   --mem-fraction-static 0.8 \
   --model-path /work/models \
   --tp-size 8 \
   --pp-size 2\
   --dp-size 1 \
   --enable-dp-attention \
   --disaggregation-mode prefill \
   --context-length 131072 \
   --disaggregation-ib-device mlx5_bond_0,mlx5_bond_1,mlx5_bond_2,mlx5_bond_3 \
   --host 0.0.0.0 \
   --port 30000 \
   --nnodes 2 \
   --node-rank 0 \
   --dist-init-addr 26.5.27.239:20000 \
   --max-running-requests 256  &
   
```
decode:
```
 SGLANG_DSV4_FP4_EXPERTS=0 
 SGLANG_JIT_DEEPGEMM_PRECOMPILE=0 
nohup python -m sglang.launch_server --model-path /work/models/ --port 30000 --trust-remote --host 0.0.0.0 --served-model-name xopglm51 --enable-cache-report --prefill-round-robin-balance --moe-a2a-backend deepep --enable-dp-lm-head  --chunked-prefill-size 131072 --enable-dp-attention --tp-size 16 --dp-size 16 --deepep-mode low_latency --context-length 131072 --mem-fraction-static 0.8 --cuda-graph-max-bs 64 --max-running-requests 1024 --ep-num-redundant-experts 16 --moe-dense-tp-size 1 --disaggregation-ib-device mlx5_bond_0,mlx5_bond_1,mlx5_bond_2,mlx5_bond_3  --allow-auto-truncate --load-balance-method total_tokens --enable-metrics --tokenizer-worker-num 8 --disaggregation-mode decode --nnodes 2 --dist-init-addr 26.5.35.131:20102 --node-rank 0 &

```
and gsm8k:
image

CC  @ShangmingCai  @fzyzcjy @Fridge003

### CODE REVIEW

File: python/sglang/srt/model_executor/input_buffers.py [hunk 1]
```diff
@@ -79,8 +80,9 @@ def create(
             )
 
             if pp_size > 1:
+                hs = pp_hidden_size if pp_hidden_size is not None else hidden_size
                 pp_proxy_tensors = {
-                    "hidden_states": torch.zeros((max_bs, hidden_size), dtype=dtype),
+                    "hidden_states": torch.zeros((max_bs, hs), dtype=dtype),
                     "residual": torch.zeros((max_bs, hidden_size), dtype=dtype),
                 }
```
Comment (ShangmingCai) [inline python/sglang/srt/model_executor/input_buffers.py:91]: Maybe need to clarify what the difference is between pp hidden size and hidden size
